realtime/redisPublisher.js

```diff
@@ -0,0 +1,134 @@
+ * createdAt timestamp, returns an update event otherwise.
+ */
+function getSampleEventType(sample) {
+  const updatedAt = new Date(sample.updatedAt).getTime();
+  const createdAt = new Date(sample.createdAt).getTime();
+  return updatedAt === createdAt ? sampleEvent.add : sampleEvent.upd;
+} // getSampleEventType
+
+/**
+ * This function returns an object to be published via redis channel
+ * @param  {Object} inst  -  Model instance
+ * @param  {[Array]} changedKeys - An array containing the fields of the model
+ * that were changed
+ * @param  {[Array]} ignoreAttributes An array containing the fields of the
+ * model that should be ignored
+ * @returns {Object} - Returns an object that is ready to be published Or null
+ * if there are no changedfields.
+ */
+function prepareToPublish(inst, changedKeys, ignoreAttributes) {
+  // prepare the data iff changed fields are not in ignoreAttributes
+  const ignoreSet = new Set(ignoreAttributes);
+  for (let i = 0; i < changedKeys.length; i++) {
+    if (!ignoreSet.has(changedKeys[i])) {
+      return {
+        old: inst, // done to maintain the shape of the object to publish
+        new: inst,
+      };
+    }
+  }
+
+  return null;
+} // prepareToPublish
+
+/**
+ * This function publishes an created, updated or a deleted model instance to
+ * the redis channel and returns the object that was published
+ *
+ * @param  {Object} inst - Model instance to be published
+ * @param  {String} event  - Type of the event that is being published
+ * @param  {[Array]} changedKeys - An array containing the fields of the model
+ * that were changed
+ * @param  {[Array]} ignoreAttributes An array containing the fields of the
+ * model that should be ignored
+ * @returns {Object} - object that was published
+ */
+function publishObject(inst, event, changedKeys, ignoreAttributes) {
+  const obj = {};
+  obj[event] = inst;
+
+  /**
+   * The shape of the object required for update events are a bit different.
+   * changedKeys and ignoreAttributes are passed in as arrays by the
+   * afterUpdate hooks of the models, which are passed to the prepareToPublish
+   * to get the object just for update events.
+   */
+  if (Array.isArray(changedKeys) && Array.isArray(ignoreAttributes)) {
+    obj[event] = prepareToPublish(inst, changedKeys, ignoreAttributes);
+  }
+
+  if (obj[event]) {
+    return pub.publish(channelName.toString(), JSON.stringify(obj));
+  }
+
+  return obj;
+} // publishChange
+
+/**
+ * The sample object needs to be attached its subject object and it also needs
+ * a absolutePath field added to it before the sample is published to the redis
+ * channel.
+ * @param  {Object} sampleInst - The sample instance to be published
+ * @param  {Model} model - The subject model used to get the related subject
+ * instance
+ * @param  {String} event  - Type of the event that is being published
+ * @returns {Promise} - which resolves to a sample object
+ */
+function publishSample(sampleInst, model, event) {
```

Comment:
We should avoid requiring files as much as possible. Requiring the subject model from the realtime module to db module might  help with not having to deal with circular dependencies in the future.